src/clj/puppetlabs/services/request_handler/request_handler_core.clj

```diff
@@ -12,6 +12,14 @@
             [slingshot.slingshot :as sling]))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Constants
+
+(def error-message-404
+  (str "Error: This master is running Puppet 4, and the URL format has changed. "
+       "This request was likely made with a Puppet 3 Agent, as it used an invalid "
+       "URL format. Please upgrade your agents to Puppet 4 and try again."))
```

Comment:
So, the biggest thing I'd like to get in here is a block of server info, listing the software and API versions involved. That's probably the nicest and most useful 404 API response I can imagine. (BTW, do we have an API endpoint that lists this info?) 

Anyway, I suggest something like this:

    Error: Invalid URL - Puppet Server expects requests that conform to the
    /puppet and /puppet-ca APIs.
    
    Note that Puppet 3 agents aren't compatible with this version; if you're
    running Puppet 3, you must either upgrade your agents to match the server
    or point them to a server running Puppet 3.

    Server info:
      Puppet Server version <VERSION>
      Puppet version <VERSION>
      Supported /puppet API versions:
        v3
      Supported /puppet-ca API versions:
        v1